ts/pubsub.js

```diff
@@ -284,9 +283,10 @@ class BasicPubSub extends Pubsub {
    * Subscribes to topics
    * @override
    * @param {Array<string>|string} topics
+   * @param {function} [handler]
    * @returns {void}
    */
-  subscribe (topics) {
+  subscribe (topics, handler) {
```

Comment:
Why attaching an event handler here?

Reply:
we want to remove this abstraction file completely from libp2p (https://github.com/libp2p/js-libp2p/blob/master/src/pubsub.js). The only thing we really needed was the handlers, as all the rest was basically protection from the APIs per https://github.com/libp2p/js-libp2p-pubsub/issues/68

I am still thinking if we can move that to the interface, but with the subscriptions I am not sure we will be able to abstract it there

Reply:
:+1:

RE: moving it to the interface:
We could do that depending on what we 'allow' to be overridable.
It seems that we only need to have routers override `_subscribe(topics: string[]): void` and `_unsubscribe(topics: string[]): void` (which MUST call super). And not `subscribe`/`unsubscribe`
In which case we could perform the event handling on `subscribe` and `unsubscribe` (before/after `_subscribe`/`_unsubscribe` is called).

Reply:
Unrelated, but I think the interface should only allow subscription/unsubscription of a single topic at a time.
eg: `subscribe(topic: string, handler: (msg: InMessage) => void): void`

Attaching a handler to multiple topics at once feels a lot like an anti-pattern, or at least a less primary interface, in the same way we were previously publishing multiple topics and multiple messages at once.

Reply:
That is a good point. Please create that discussion in the interface PR